plugins/SlicerT/SlicerTWaveform.cpp

```diff
@@ -0,0 +1,418 @@
+/*
+ * SlicerTWaveform.cpp - slice editor for SlicerT
+ *
+ * Copyright (c) 2023 Daniel Kauss Serna <daniel.kauss.serna@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SlicerTWaveform.h"
+
+#include <QBitmap>
+
+#include "SlicerT.h"
+#include "SlicerTView.h"
+#include "embed.h"
+
+namespace lmms {
+
+namespace gui {
+
+SlicerTWaveform::SlicerTWaveform(int w, int h, SlicerT* instrument, QWidget* parent)
+	: QWidget(parent)
+	// calculate sizes
+	, m_width(w)
+	, m_height(h)
+	, m_seekerWidth(w - m_seekerHorMargin * 2)
+	, m_editorHeight(h - m_seekerHeight - m_middleMargin)
+	, m_editorWidth(w)
+
+	// create pixmaps
+	, m_sliceArrow(PLUGIN_NAME::getIconPixmap("slice_indicator_arrow"))
+	, m_seeker(QPixmap(m_seekerWidth, m_seekerHeight))
+	, m_seekerWaveform(QPixmap(m_seekerWidth, m_seekerHeight))
+	, m_editorWaveform(QPixmap(m_editorWidth, m_editorHeight))
+	, m_sliceEditor(QPixmap(w, m_editorHeight))
+	, m_emptySampleIcon(embed::getIconPixmap("sample_track.png"))
+
+	// references to instrument vars
+	, m_slicerTParent(instrument)
+{
+	// window config
+	setFixedSize(m_width, m_height);
+	setMouseTracking(true);
+
+	// draw backgrounds
+	m_seekerWaveform.fill(s_waveformBgColor);
+	m_editorWaveform.fill(s_waveformBgColor);
+
+	// connect to playback
+	connect(instrument, &SlicerT::isPlaying, this, &SlicerTWaveform::isPlaying);
+	connect(instrument, &SlicerT::dataChanged, this, &SlicerTWaveform::updateUI);
+
+	// preprocess icons
+	m_emptySampleIcon = m_emptySampleIcon.createMaskFromColor(QColor(255, 255, 255), Qt::MaskMode::MaskOutColor);
+
+	updateUI();
+}
+
+void SlicerTWaveform::drawSeekerWaveform()
+{
+	m_seekerWaveform.fill(s_waveformBgColor);
+	if (m_slicerTParent->m_originalSample.frames() < 2048) { return; }
```

Comment:
> Cant we just prevent loading of samples too small before sending it to the waveform and other objects? What I mean is that wherever you have your loadSample function or whatnot, just prevent loading it from there and maybe put up a warning dialog using QMessageBox::warning.

These checks are actually for UI interactions. If you click a button, it checks if the sample is valid. That's why it is not enough to not load. Since I removed time shifting, there should be no lower limit to sample size, so I changed all 2048 to 1.